From a0a2310c05b8205b67a9e4aeaa6bb98d68746b52 Mon Sep 17 00:00:00 2001
From: Dmitry Timoshkov <dmitry@baikal.ru>
Date: Sat, 31 Oct 2015 14:31:51 +0800
Subject: widl: Avoid generating duplicate typelib entries for structure tag
 names. Resend.

I've added the tests to show how it's supposed to behave. Without this
patch a generated typelib has numerous duplicate '_n' entries.

This is merely an optimization for the typelib generator, but this change
breaks header generation, thus it's limited only to the typelib mode.

Signed-off-by: Dmitry Timoshkov <dmitry@baikal.ru>
---
 tools/widl/typetree.c | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/tools/widl/typetree.c b/tools/widl/typetree.c
index 5925d40..e316614 100644
--- a/tools/widl/typetree.c
+++ b/tools/widl/typetree.c
@@ -300,7 +300,12 @@ type_t *type_new_enum(const char *name, struct namespace *namespace, int defined
 type_t *type_new_struct(char *name, struct namespace *namespace, int defined, var_list_t *fields)
 {
     type_t *tag_type = name ? find_type(name, namespace, tsSTRUCT) : NULL;
-    type_t *t = make_type(TYPE_STRUCT);
+    type_t *t;
+
+    /* avoid creating duplicate typelib type entries */
+    if (tag_type && do_typelib) return tag_type;
+
+    t = make_type(TYPE_STRUCT);
     t->name = name;
     t->namespace = namespace;
 
-- 
2.6.1

